crates/matrix-sdk-crypto/src/backups/keys/backup.rs

```diff
@@ -28,9 +31,14 @@ use zeroize::Zeroizing;
 use super::{compat::PkEncryption, decryption::DecodeError};
 use crate::olm::InboundGroupSession;
 
+pub type HmacSha256 = Hmac<Sha256>;
+
+pub type HmacSha256Key = Zeroizing<[u8; 32]>;
```

Comment:
We usually use `Box<[u8; 32]>` for this. And add derive `Zeroize` and `ZeroizeOnDrop` on the container type.

Reply:
If I do that, and pass the `Box[<u8; 32]>` as a function argument, as in `MegolmV1BackupKey::new`, then it will automagically do the right thing when it becomes part of the struct?

Reply:
If by right thing you mean that it will zeroize the `Box<[u8; 32]>` when the whole struct gets dropped, then yes.

Reply:
I tried deriving `Zeroize` and `ZeroizeOnDrop` on the container type, which is `InnerBackupKey`

```rust
pub type HmacSha256Key = Box<[u8; 32]>;

#[derive(Debug, Zeroize, ZeroizeOnDrop)]
struct InnerBackupKey {
    key: Curve25519PublicKey,
    mac_key: Option<HmacSha256Key>,
    signatures: BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, String>>,
    version: Mutex<Option<String>>,
}
```

but that gives me errors:

```
error[E0599]: the method `zeroize` exists for mutable reference `&mut Curve25519PublicKey`, but its trait bounds were not satisfied
   --> crates/matrix-sdk-crypto/src/backups/keys/backup.rs:38:17
    |
38  | #[derive(Debug, Zeroize, ZeroizeOnDrop)]
    |                 ^^^^^^^ method cannot be called on `&mut Curve25519PublicKey` due to unsatisfied trait bounds
    |
   ::: /home/hubert/.cargo/registry/src/index.crates.io-6f17d22bba15001f/vodozemac-0.5.0/src/types/curve25519.rs:110:1
    |
110 | pub struct Curve25519PublicKey {
    | ------------------------------
    | |
    | doesn't satisfy `Curve25519PublicKey: DefaultIsZeroes`
    | doesn't satisfy `Curve25519PublicKey: Zeroize`
    |
    = note: the following trait bounds were not satisfied:
            `Curve25519PublicKey: DefaultIsZeroes`
            which is required by `Curve25519PublicKey: Zeroize`
            `&mut Curve25519PublicKey: DefaultIsZeroes`
            which is required by `&mut Curve25519PublicKey: Zeroize`
    = note: this error originates in the derive macro `Zeroize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `zeroize` exists for mutable reference `&mut Option<Box<[u8; 32]>>`, but its trait bounds were not satisfied
  --> crates/matrix-sdk-crypto/src/backups/keys/backup.rs:38:17
   |
38 | #[derive(Debug, Zeroize, ZeroizeOnDrop)]
   |                 ^^^^^^^ method cannot be called on `&mut Option<Box<[u8; 32]>>` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `&mut std::option::Option<Box<[u8; 32]>>: DefaultIsZeroes`
           which is required by `&mut std::option::Option<Box<[u8; 32]>>: Zeroize`
   = note: this error originates in the derive macro `Zeroize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `zeroize` exists for mutable reference `&mut BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, String>>`, but its trait bounds were not satisfied
   --> crates/matrix-sdk-crypto/src/backups/keys/backup.rs:38:17
    |
38  |   #[derive(Debug, Zeroize, ZeroizeOnDrop)]
    |                   ^^^^^^^ method cannot be called due to unsatisfied trait bounds
    |
   ::: /home/hubert/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/collections/btree/map.rs:172:1
    |
172 | / pub struct BTreeMap<
173 | |     K,
174 | |     V,
175 | |     #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator + Clone = Global,
176 | | > {
    | | -
    | | |
    | |_doesn't satisfy `_: DefaultIsZeroes`
    |   doesn't satisfy `_: Zeroize`
    |
    = note: the following trait bounds were not satisfied:
            `BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, std::string::String>>: DefaultIsZeroes`
            which is required by `BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, std::string::String>>: Zeroize`
            `&mut BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, std::string::String>>: DefaultIsZeroes`
            which is required by `&mut BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, std::string::String>>: Zeroize`
    = note: this error originates in the derive macro `Zeroize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `zeroize` exists for mutable reference `&mut Mutex<Option<String>>`, but its trait bounds were not satisfied
   --> crates/matrix-sdk-crypto/src/backups/keys/backup.rs:38:17
    |
38  | #[derive(Debug, Zeroize, ZeroizeOnDrop)]
    |                 ^^^^^^^ method cannot be called on `&mut Mutex<Option<String>>` due to unsatisfied trait bounds
    |
   ::: /home/hubert/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sync/mutex.rs:175:1
    |
175 | pub struct Mutex<T: ?Sized> {
    | ---------------------------
    | |
    | doesn't satisfy `_: DefaultIsZeroes`
    | doesn't satisfy `_: Zeroize`
    |
    = note: the following trait bounds were not satisfied:
            `std::sync::Mutex<std::option::Option<std::string::String>>: DefaultIsZeroes`
            which is required by `std::sync::Mutex<std::option::Option<std::string::String>>: Zeroize`
            `&mut std::sync::Mutex<std::option::Option<std::string::String>>: DefaultIsZeroes`
            which is required by `&mut std::sync::Mutex<std::option::Option<std::string::String>>: Zeroize`
    = note: this error originates in the derive macro `Zeroize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `zeroize_or_on_drop` exists for mutable reference `&mut Curve25519PublicKey`, but its trait bounds were not satisfied
   --> crates/matrix-sdk-crypto/src/backups/keys/backup.rs:38:26
    |
38  | #[derive(Debug, Zeroize, ZeroizeOnDrop)]
    |                          ^^^^^^^^^^^^^ method cannot be called on `&mut Curve25519PublicKey` due to unsatisfied trait bounds
    |
   ::: /home/hubert/.cargo/registry/src/index.crates.io-6f17d22bba15001f/vodozemac-0.5.0/src/types/curve25519.rs:110:1
    |
110 | pub struct Curve25519PublicKey {
    | ------------------------------
    | |
    | doesn't satisfy `Curve25519PublicKey: AssertZeroize`
    | doesn't satisfy `Curve25519PublicKey: ZeroizeOnDrop`
    | doesn't satisfy `Curve25519PublicKey: Zeroize`
    |
    = note: the following trait bounds were not satisfied:
            `Curve25519PublicKey: Zeroize`
            which is required by `Curve25519PublicKey: AssertZeroize`
            `Curve25519PublicKey: ZeroizeOnDrop`
            which is required by `&&mut Curve25519PublicKey: AssertZeroizeOnDrop`
            `&mut Curve25519PublicKey: Zeroize`
            which is required by `&mut Curve25519PublicKey: AssertZeroize`
    = note: this error originates in the derive macro `ZeroizeOnDrop` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `zeroize_or_on_drop` exists for mutable reference `&mut Option<Box<[u8; 32]>>`, but its trait bounds were not satisfied
   --> crates/matrix-sdk-crypto/src/backups/keys/backup.rs:38:26
    |
38  | #[derive(Debug, Zeroize, ZeroizeOnDrop)]
    |                          ^^^^^^^^^^^^^ method cannot be called on `&mut Option<Box<[u8; 32]>>` due to unsatisfied trait bounds
    |
   ::: /home/hubert/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:563:1
    |
563 | pub enum Option<T> {
    | ------------------
    | |
    | doesn't satisfy `std::option::Option<Box<[u8; 32]>>: AssertZeroize`
    | doesn't satisfy `std::option::Option<Box<[u8; 32]>>: ZeroizeOnDrop`
    | doesn't satisfy `std::option::Option<Box<[u8; 32]>>: Zeroize`
    |
    = note: the following trait bounds were not satisfied:
            `std::option::Option<Box<[u8; 32]>>: Zeroize`
            which is required by `std::option::Option<Box<[u8; 32]>>: AssertZeroize`
            `std::option::Option<Box<[u8; 32]>>: ZeroizeOnDrop`
            which is required by `&&mut std::option::Option<Box<[u8; 32]>>: AssertZeroizeOnDrop`
            `&mut std::option::Option<Box<[u8; 32]>>: Zeroize`
            which is required by `&mut std::option::Option<Box<[u8; 32]>>: AssertZeroize`
    = note: this error originates in the derive macro `ZeroizeOnDrop` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `zeroize_or_on_drop` exists for mutable reference `&mut BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, String>>`, but its trait bounds were not satisfied
   --> crates/matrix-sdk-crypto/src/backups/keys/backup.rs:38:26
    |
38  |   #[derive(Debug, Zeroize, ZeroizeOnDrop)]
    |                            ^^^^^^^^^^^^^ method cannot be called due to unsatisfied trait bounds
    |
   ::: /home/hubert/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/collections/btree/map.rs:172:1
    |
172 | / pub struct BTreeMap<
173 | |     K,
174 | |     V,
175 | |     #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator + Clone = Global,
176 | | > {
    | | -
    | | |
    | | doesn't satisfy `_: AssertZeroize`
    | |_doesn't satisfy `_: ZeroizeOnDrop`
    |   doesn't satisfy `_: Zeroize`
    |
    = note: the following trait bounds were not satisfied:
            `BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, std::string::String>>: Zeroize`
            which is required by `BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, std::string::String>>: AssertZeroize`
            `BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, std::string::String>>: ZeroizeOnDrop`
            which is required by `&&mut BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, std::string::String>>: AssertZeroizeOnDrop`
            `&mut BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, std::string::String>>: Zeroize`
            which is required by `&mut BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, std::string::String>>: AssertZeroize`
    = note: this error originates in the derive macro `ZeroizeOnDrop` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `zeroize_or_on_drop` exists for mutable reference `&mut Mutex<Option<String>>`, but its trait bounds were not satisfied
   --> crates/matrix-sdk-crypto/src/backups/keys/backup.rs:38:26
    |
38  | #[derive(Debug, Zeroize, ZeroizeOnDrop)]
    |                          ^^^^^^^^^^^^^ method cannot be called on `&mut Mutex<Option<String>>` due to unsatisfied trait bounds
    |
   ::: /home/hubert/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sync/mutex.rs:175:1
    |
175 | pub struct Mutex<T: ?Sized> {
    | ---------------------------
    | |
    | doesn't satisfy `_: AssertZeroize`
    | doesn't satisfy `_: ZeroizeOnDrop`
    | doesn't satisfy `_: Zeroize`
    |
    = note: the following trait bounds were not satisfied:
            `std::sync::Mutex<std::option::Option<std::string::String>>: Zeroize`
            which is required by `std::sync::Mutex<std::option::Option<std::string::String>>: AssertZeroize`
            `std::sync::Mutex<std::option::Option<std::string::String>>: ZeroizeOnDrop`
            which is required by `&&mut std::sync::Mutex<std::option::Option<std::string::String>>: AssertZeroizeOnDrop`
            `&mut std::sync::Mutex<std::option::Option<std::string::String>>: Zeroize`
            which is required by `&mut std::sync::Mutex<std::option::Option<std::string::String>>: AssertZeroize`
    = note: this error originates in the derive macro `ZeroizeOnDrop` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0599`.
```

Should I do something else, or is there a way to fix this?

Reply:
The error complains that not all the types in the fields of the struct implement `Zeroize`. But this is completely fine since only one field contains private key material which should be zeroized.

You can make the compiler happy by implementing `Zeroize` manually, or use the `#[zeroize(skip)` attribute on the fields which don't need to be zeroized, more info about this macro can be found in the `zeroize` crate docs: https://docs.rs/zeroize/latest/zeroize/index.html#custom-derive-support

Reply:
I did

```rust
pub type HmacSha256Key = Box<[u8; 32]>;

#[derive(Debug, Zeroize, ZeroizeOnDrop)]
struct InnerBackupKey {
    #[zeroize(skip)]
    key: Curve25519PublicKey,
    mac_key: Option<HmacSha256Key>,
    #[zeroize(skip)]
    signatures: Signatures,
    #[zeroize(skip)]
    version: Mutex<Option<String>>,
}
```

but I get

```
error[E0599]: the method `zeroize` exists for mutable reference `&mut Option<Box<[u8; 32]>>`, but its trait bounds were not satisfied
  --> crates/matrix-sdk-crypto/src/backups/keys/backup.rs:34:17
   |
34 | #[derive(Debug, Zeroize, ZeroizeOnDrop)]
   |                 ^^^^^^^ method cannot be called on `&mut Option<Box<[u8; 32]>>` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `&mut std::option::Option<Box<[u8; 32]>>: DefaultIsZeroes`
           which is required by `&mut std::option::Option<Box<[u8; 32]>>: Zeroize`
   = note: this error originates in the derive macro `Zeroize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `zeroize_or_on_drop` exists for mutable reference `&mut Option<Box<[u8; 32]>>`, but its trait bounds were not satisfied
   --> crates/matrix-sdk-crypto/src/backups/keys/backup.rs:34:26
    |
34  | #[derive(Debug, Zeroize, ZeroizeOnDrop)]
    |                          ^^^^^^^^^^^^^ method cannot be called on `&mut Option<Box<[u8; 32]>>` due to unsatisfied trait bounds
    |
   ::: /home/hubert/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:563:1
    |
563 | pub enum Option<T> {
    | ------------------
    | |
    | doesn't satisfy `std::option::Option<Box<[u8; 32]>>: AssertZeroize`
    | doesn't satisfy `std::option::Option<Box<[u8; 32]>>: ZeroizeOnDrop`
    | doesn't satisfy `std::option::Option<Box<[u8; 32]>>: Zeroize`
    |
    = note: the following trait bounds were not satisfied:
            `std::option::Option<Box<[u8; 32]>>: Zeroize`
            which is required by `std::option::Option<Box<[u8; 32]>>: AssertZeroize`
            `std::option::Option<Box<[u8; 32]>>: ZeroizeOnDrop`
            which is required by `&&mut std::option::Option<Box<[u8; 32]>>: AssertZeroizeOnDrop`
            `&mut std::option::Option<Box<[u8; 32]>>: Zeroize`
            which is required by `&mut std::option::Option<Box<[u8; 32]>>: AssertZeroize`
    = note: this error originates in the derive macro `ZeroizeOnDrop` (in Nightly builds, run with -Z macro-backtrace for more info)
```

So it seems like zeroize doesn't like `Option`.  One solution that I can think of is to make `mac_key: HmacSha256Key`, so it always has some sort of key stored, but add a boolean flag that indicates whether the key is actually used.  But that doesn't seem very Rust-y to me.  Are there any better solutions?

Reply:
It's not the `Option<T>` per se, it's the `Option<Box<T>>`, not quite sure why that would be since `zeroizing` implements `Zeroize` for both `Option<T>` and `Box<T>`. Alright, let's stick to the `Zeroizing` thing since this seems to be too annoying.